ts/components/session/conversation/SessionConversation.tsx

```diff
@@ -0,0 +1,860 @@
+
+    const scrollButtonViewShowLimit = 0.75;
+    const scrollButtonViewHideLimit = 0.40;
+    const scrollOffsetPx = scrollHeight - scrollTop - clientHeight;
+    const scrollOffsetPc = scrollOffsetPx / clientHeight;
+
+    // Scroll button appears if you're more than 75% scrolled up
+    if (scrollOffsetPc > scrollButtonViewShowLimit && !this.state.showScrollButton){
+      this.setState({showScrollButton: true});
+    }
+    // Scroll button disappears if you're more less than 40% scrolled up
+    if (scrollOffsetPc < scrollButtonViewHideLimit && this.state.showScrollButton){
+      this.setState({showScrollButton: false});
+    }
+
+    console.log(`[scroll] scrollOffsetPx: `, scrollOffsetPx);
+    console.log(`[scroll] scrollOffsetPc: `, scrollOffsetPc);
+
+    // Scrolled to bottom
+    const isScrolledToBottom = scrollOffsetPc === 0;
+    if (isScrolledToBottom) {
+      console.log(`[scroll] Scrolled to bottom`);
+    }
+
+    // Mark messages read
+    this.updateReadMessages();
+
+    // Pin scroll to bottom on new message, unless user has scrolled up
+    if (this.state.isScrolledToBottom !== isScrolledToBottom){
+      this.setState({ isScrolledToBottom });
+    }
+
+    // Fetch more messages when nearing the top of the message list
+    const shouldFetchMoreMessages = scrollTop <= window.CONSTANTS.MESSAGE_CONTAINER_BUFFER_OFFSET_PX;
+
+    if (shouldFetchMoreMessages){
+      const numMessages = this.state.messages.length + window.CONSTANTS.DEFAULT_MESSAGE_FETCH_COUNT;
+
+      // Prevent grabbing messags with scroll more frequently than once per 5s.
+      const messageFetchInterval = 2;
+      const previousTopMessage = (await this.getMessages(numMessages, messageFetchInterval, true))?.previousTopMessage;
+      previousTopMessage && this.scrollToMessage(previousTopMessage);
+    }
+  }
+
+  public scrollToUnread() {
+    const { messages, unreadCount } = this.state;
+    const message = messages[(messages.length - 1) - unreadCount];
+
+    if (message) {
+      this.scrollToMessage(message.id);
+    }
+  }
+
+  public scrollToMessage(messageId: string) {
+    const topUnreadMessage = document.getElementById(messageId);
+    topUnreadMessage?.scrollIntoView();
+  }
+
+  public scrollToBottom() {
+    // FIXME VINCE: Smooth scrolling that isn't slow@!
+    // this.messagesEndRef.current?.scrollIntoView(
+    //   { behavior: firstLoad ? 'auto' : 'smooth' }
+    // );
+
+    const messageContainer = this.messageContainerRef.current;
+    if (!messageContainer) {
+      return;
+    }
+    messageContainer.scrollTop = messageContainer.scrollHeight - messageContainer.clientHeight;
+  }
+
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // ~~~~~~~~~~~~ MESSAGE SELECTION ~~~~~~~~~~~~~
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  public selectMessage(messageId: string) {
+    const selectedMessages = this.state.selectedMessages.includes(messageId)
+      // Add to array if not selected. Else remove.
```

Comment:
Comment should be `// Remove from array if selected else add it` so it matches the code